driver/gather_driver.hpp

```diff
@@ -0,0 +1,389 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include "InputFlags.hpp"
+#include "driver.hpp"
+#include "mloGatherHost.hpp"
+#include "random.hpp"
+#include "tensor_driver.hpp"
+#include "timer.hpp"
+
+#include <../test/verify.hpp>
+
+#include <miopen/errors.hpp>
+#include <miopen/miopen.h>
+#include <miopen/tensor_view_utils.hpp>
+
+#include <memory>
+#include <vector>
+
+template <typename Tgpu, typename Tref, typename Tindex>
```

Comment:
Can `Tindex` be any datatype different than `int64_t`? If not, then let's fixed it to be `int64_t` instead of create a separate instance `Tindex` as a template.

Btw, can `index` be negative? I dont think MIOpen support for negative index, so I think it should be `uint64_t` instead. pls check it again.

Reply:
- Can `Tindex` be any datatype different than `int64_t`: no if only for Gather op. This driver is intended for general gather-related ops (gather, gathernd, gatherv2) and 2 of them support different `Tindex` types
- Can `index` be negative: no, I've updated it to use `uint64_t` as you suggested

Reply:
I see. 

Just one more additional question, what are other different `Tindex` types that other gather-related op support? Can you list them here? 

I'm just a little confused as I thought we should set all index-related tensor to types of `uint64_t`, I'm uncertain about the rationale for using different data types for tensors that contain index-referencing elements.

Reply:
`Tindex` can be `uint32_t` or `uint64_t` (maybe in pytorch, it mainly supports `uint64_t` as type of index tensor but tensorflow also support `int32_t`)